package org.ncgr.pangenomics.genotype;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;

import java.util.Arrays;
import java.util.List;
import java.util.LinkedList;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.TreeMap;
import java.util.Set;
import java.util.HashSet;
import java.util.TreeSet;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import org.jgrapht.graph.DirectedAcyclicGraph;

import org.mskcc.cbio.portal.stats.FisherExact;

public class PangenomicGraph extends DirectedAcyclicGraph<Node,Edge> {

    // output verbosity
    public boolean verbose = false;

    // a name for this graph
    public String name;

    // Set of Paths that traverse this graph
    public TreeSet<Path> paths = new TreeSet<>();

    // Map of path.getName() to the Path (convenience)
    public TreeMap<String,Path> pathNameMap = new TreeMap<>();

    // Map of Node ids to their Nodes (convenience)
    public TreeMap<Long,Node> nodeIdMap = new TreeMap<>();

    // maps a Node to the Paths that traverse it; private so that we're forced to build it if it's empty
    private TreeMap<Node,List<Path>> nodePaths = new TreeMap<>();

    // this graph's samples
    public TreeSet<Sample> samples = new TreeSet<>();
    
    // maps a path label to a count of paths that have that label
    public TreeMap<String,Integer> labelCounts = new TreeMap<>(); // keyed by label, construct in method

    // computed once to save time
    public FisherExact fisherExact;

    /**
     * Basic constructor.
     */
    public PangenomicGraph(String name) {
        super(Edge.class);
	this.name = name;
    }

    /**
     * Build this graph's paths from Maps generated by an Importer and filters.
     */
    public void buildGraphPaths(TreeMap<Sample,NodeSet> sampleNodeSets, TreeMap<Node,TreeSet<Sample>> nodeSamples, int maxCases, boolean equalizeCasesControls) {
	// load samples from the sampleNodeSets
	samples = new TreeSet<>(sampleNodeSets.keySet());
	// FILTER: limit maximum cases
	if (maxCases>0) {
	    int nCases = 0;
	    int nControls = 0;
            for (Sample sample : samples) {
                if (sample.isCase()) {
                    nCases++;
                } else if (sample.isControl()) {
                    nControls++;
                }
            }
            Set<Sample> samplesToDrop = new HashSet<>();
            int numberToRemove = nCases - maxCases;
	    int count = 0;
	    // randomly select samples to drop
	    while (count<numberToRemove) {
		Optional<Sample> optional = samples.stream().skip((int)(samples.size()*Math.random())).findFirst();
		if (optional.isPresent()) {
		    Sample sample = optional.get();
		    if (count<numberToRemove && sample.isCase() && !samplesToDrop.contains(sample)) {
			samplesToDrop.add(sample);
			count++;
		    }
                }
	    }
	    // remove dropped samples
	    for (Sample sample : samplesToDrop) {
		samples.remove(sample);
		sampleNodeSets.remove(sample);
		for (Node n : nodeSamples.keySet()) {
		    Set<Sample> samples = nodeSamples.get(n);
		    samples.remove(sample);
		}
            }
            // drop nodes that no longer have any samples
            Set<Node> nodesToDrop = new HashSet<>();
            for (Node n : nodeSamples.keySet()) {
                Set<Sample> samples = nodeSamples.get(n);
                if (samples.size()==0) nodesToDrop.add(n);
            }
	    for (Node n : nodesToDrop) {
		nodeIdMap.remove(n.id);
		nodeSamples.remove(n);
	    }
            System.err.println("Removed "+samplesToDrop.size()+" samples and "+nodesToDrop.size()+" orphaned nodes to set number of cases = "+maxCases+".");
	}
	// FILTER: enforce cases=controls
        if (equalizeCasesControls) {
            int nCases = 0;
            int nControls = 0;
            for (Sample sample : samples) {
                if (sample.isCase()) {
                    nCases++;
                } else if (sample.isControl()) {
                    nControls++;
                }
            }
            Set<Sample> samplesToDrop = new HashSet<>();
            int numberToRemove = Math.abs(nCases - nControls);
            int count = 0;
            // randomly select samples to drop
            while (count<numberToRemove) {
                Optional<Sample> optional = samples.stream().skip((int)(samples.size()*Math.random())).findFirst();
                if (optional.isPresent()) {
                    Sample sample = optional.get();
                    if (nCases>nControls && sample.isCase() && !samplesToDrop.contains(sample)) {
                        samplesToDrop.add(sample);
                        count++;
                    } else if (nControls>nCases && sample.isControl() && !samplesToDrop.contains(sample)) {
                        samplesToDrop.add(sample);
                        count++;
                    }
                }
	    }
	    // remove dropped samples
	    for (Sample sample : samplesToDrop) {
		samples.remove(sample);
		sampleNodeSets.remove(sample);
		for (Node n : nodeSamples.keySet()) {
		    Set<Sample> samples = nodeSamples.get(n);
		    samples.remove(sample);
		}
            }
            // drop nodes that no longer have any samples
            Set<Node> nodesToDrop = new HashSet<>();
            for (Node n : nodeSamples.keySet()) {
                Set<Sample> samples = nodeSamples.get(n);
                if (samples.size()==0) nodesToDrop.add(n);
            }
	    for (Node n : nodesToDrop) {
		nodeIdMap.remove(n.id);
		nodeSamples.remove(n);
	    }
            System.err.println("Removed "+samplesToDrop.size()+" samples and "+nodesToDrop.size()+" orphaned nodes to equalize cases and controls.");
        }
        // build the paths and path-labeled graph edges from the samples and sampleNodeSets map
        if (verbose) System.err.println("Creating paths and adding edges to graph...");
        for (Sample sample : samples) {
	    if (sampleNodeSets.containsKey(sample)) {
		NodeSet sampleNodes = sampleNodeSets.get(sample);
		Path path = new Path(this, new LinkedList<Node>(sampleNodes), sample);
		paths.add(path);
		pathNameMap.put(sample.name, path);
		// add edges
		Node lastNode = null;
		for (Node sampleNode : sampleNodes) {
		    // sampleNode only contains id -- get actual node from graph for this id
		    Node node = getNode(sampleNode.id);
		    if (lastNode!=null) {
			if (!containsEdge(lastNode, node)) {
			    try {
				addEdge(lastNode, node);
			    } catch (Exception e) {
				System.err.println("ERROR adding edge from "+lastNode+" to "+node);
				System.err.println("from node="+lastNode);
				System.err.println("to node="+node);
				System.err.println(e);
				System.exit(1);
			    }
			}
		    }
		    lastNode = node;
		}
	    }
        }
	// initialize FisherExact for later use
        fisherExact = new FisherExact(paths.size());
    }
    
    /**
     * Build this graph from the provided Nodes and Paths.
     */
    public void buildGraph(TreeMap<Long,Node> nodes, TreeSet<Path> paths) {
        this.paths = paths;
        // add the nodes as graph vertices
        if (verbose) System.err.println("Adding nodes to graph vertices...");
        for (Node n : nodes.values()) {
            addVertex(n);
        }
        // build the path-labeled graph edges
        if (verbose) System.err.println("Adding edges to graph from paths...");
        for (Path path : paths) {
            pathNameMap.put(path.getName(), path);
            // add edges
            Node lastNode = null;
            for (Object n : path.getNodes()) {
                Node node = (Node) n;
                if (lastNode!=null) {
                    if (!containsEdge(lastNode, node)) {
                        try {
                            addEdge(lastNode, node);
                        } catch (Exception e) {
                            System.err.println("ERROR adding edge from "+lastNode+" to "+node);
                            System.err.println(e);
                            System.exit(1);
                        }
                    }
                }
                lastNode = node;
            }
        }
        fisherExact = new FisherExact(paths.size());
    }
    
    /**
     * Build this graph's nodes from the provided Nodes map.
     */
    public void buildGraphNodes(TreeMap<Long,Node> nodes) { // Map<String,NodeSet> sampleNodeSets, Map<Node,TreeSet<String>> nodeSamples) {
	if (nodes.size()==0) {
	    System.err.println("ERROR in PangenomicGraph.buildGraph: nodes map is empty.");
	    System.exit(1);
	}
        // add the nodes as graph vertices
        if (verbose) System.err.println("Adding nodes to graph vertices...");
        for (Node n : nodes.values()) {
            addVertex(n);
	    nodeIdMap.put(n.id, n);
        }
    }

    /**
     * Tally the label counts for the paths in this graph (can be a subset of all samples/paths).
     */
    public void tallyLabelCounts() {
	// clear just in case
	for (String label : labelCounts.keySet()) {
	    labelCounts.put(label, 0);
	}
        for (Path path : paths) {
            if (labelCounts.containsKey(path.getLabel())) {
                int count = labelCounts.get(path.getLabel());
                labelCounts.put(path.getLabel(), count+1);
            } else {
                labelCounts.put(path.getLabel(), 1);
            }
        }
    }

    /**
     * Build the node paths: the set of paths that run through each node.
     */
    public void buildNodePaths() {
	if (verbose) System.err.println("Building node paths...");
        // initialize empty paths for each node
        for (Node n : getNodes()) {
            nodePaths.put(n, new LinkedList<Path>());
        }
        // now load the paths
	for (Path path : paths) {
	    List<Node> pathNodes = path.getNodes();
	    for (Node n : pathNodes) {
		nodePaths.get(n).add(path);
	    }
	}
	// validation
	for (Node n : nodePaths.keySet()) {
	    List<Path> pathList = nodePaths.get(n);
	    for (Path p : pathList) {
		if (p==null) {
		    System.err.println("ERROR: p==null for node "+n.toString());
		    System.exit(1);
		}
	    }
	}
    }

    /**
     * Return the node paths, building them first if needed.
     */
    public TreeMap<Node,List<Path>> getNodePaths() {
	if (nodePaths.size()==0) buildNodePaths();
	return nodePaths;
    }

    /**
     * Return the node with the given id, else null.
     */
    public Node getNode(long id) {
        return nodeIdMap.get(id);
    }

    /**
     * Return true if this graph contains the path with the given name.
     */
    public boolean hasPath(String name) {
        return pathNameMap.containsKey(name);
    }
    
    /**
     * Return the path with the given name.
     */
    public Path getPath(String name) {
        return pathNameMap.get(name);
    }

    /**
     * Return the number of this graph's paths.
     */
    public int getPathCount() {
        return paths.size();
    }

    /**
     * Get the count of paths with the given label.
     */
    public int getPathCount(String label) {
        if (labelCounts.containsKey(label)) {
            return labelCounts.get(label);
        } else {
            return 0;
        }
    }

    /**
     * Return the number of paths that traverse the given node.
     */
    public int getPathCount(Node n) {
        return getPaths(n).size();
    }

    /**
     * Get the total count of paths that follow the given Edge.
     */
    public int getPathCount(Edge e) {
        int count = 0;
        for (Path p : paths) {
            List<Edge> edges = p.getEdges();
            if (edges.contains(e)) {
                count++;
            }
        }
        return count;
    }

    /**
     * Get the count of paths with the given label that follow the given Edge.
     */
    public int getPathCount(Edge e, String label) {
        int count = 0;
        for (Path p : paths) {
            if (p.getLabel().equals(label)) {
                List<Edge> edges = p.getEdges();
                if (edges.contains(e)) {
                    count++;
                }
            }
        }
        return count;
    }

    /**
     * Return the paths that traverse the given node.
     */
    public List<Path> getPaths(Node n) {
        return getNodePaths().get(n);
    }

    /**
     * Just a synonym for vertexSet(), but puts the Nodes in a List.
     */
    public List<Node> getNodes() {
        return new LinkedList<Node>(vertexSet());
    }

    /**
     * Construct a NodeSet from a string representation, e.g. [1350,1352,1353,1465,1467,1468,1469].
     */
    public NodeSet getNodeSet(String str) {
        NodeSet nodeSet = new NodeSet();
        List<Node> allNodes = getNodes();
        Map<Long,Node> allNodeMap = new HashMap<>();
        for (Node n : allNodes) allNodeMap.put(n.id, n);
        List<String> nodeStrings = Arrays.asList(str.replace("[","").replace("]","").split(","));
        for (String s : nodeStrings) {
            if (s.length()>0) {
                long id = Long.parseLong(s);
                if (allNodeMap.containsKey(id)) {
                    nodeSet.add(allNodeMap.get(id));
                } else {
                    // bail, we're asked for a node that is not in the graph
		    System.err.println("ERROR: Graph does not contain node "+id);
		    System.exit(1);
		}
            }
        }
        return nodeSet;
    }

    /**
     * Get the label counts map for paths that traverse the given node.
     */
    public Map<String,Integer> getLabelCounts(Node n) {
        Map<String,Integer> map = new HashMap<>();
        for (Path p : getNodePaths().get(n)) {
            if (map.containsKey(p.getLabel())) {
                // increment count
                int count = map.get(p.getLabel()) + 1;
                map.put(p.getLabel(), count);
            } else {
                // initialize count
                map.put(p.getLabel(), 1);
            }
        }
        return map;
    }

    /**
     * Get the label counts for paths that follow the given Edge.
     * TODO: restrict path loop to those that are on Nodes connected by Edge. May not be worth it.
     */
    public Map<String,Integer> getLabelCounts(Edge e) {
        Map<String,Integer> map = new HashMap<>();
        for (Path p : paths) {
            List<Edge> edges = p.getEdges();
            if (edges.contains(e)) {
                if (map.containsKey(p.getLabel())) {
                    // increment count
                    int count = map.get(p.getLabel()) + 1;
                    map.put(p.getLabel(), count);
                } else {
                    // initialize count
                    map.put(p.getLabel(), 1);
                }
            }
        }
        return map;
    }

    /**
     * Return the path names as a string array.
     */
    public String[] getPathNames() {
        String[] names = new String[paths.size()];
	int i = 0;
	for (Path p : paths) {
	    names[i++] = p.getName();
	}
        return names;
    }

    /**
     * Return the odds ratio for case/control paths that traverse the given node vs. case/control paths that don't.
     * 0                 = all control paths traverse node
     * POSITIVE_INFINITY = all case paths traverse node
     */
    public double oddsRatio(Node n) {
        Map<String,Integer> countsMap = getLabelCounts(n);
        int nodeCasePaths = 0; if (countsMap.containsKey("case")) nodeCasePaths = countsMap.get("case");
        int nodeCtrlPaths = 0; if (countsMap.containsKey("ctrl")) nodeCtrlPaths = countsMap.get("ctrl");
        int otherCasePaths = labelCounts.get("case") - nodeCasePaths;
        int otherCtrlPaths = labelCounts.get("ctrl") - nodeCtrlPaths;
        return (double)(nodeCasePaths*otherCtrlPaths) / (double)(nodeCtrlPaths*otherCasePaths);
    }

    /**
     * Return the Fisher's exact test two-tailed p value for case/control paths that traverse the given node
     * vs. case/control paths that don't.
     *
     *      | traverse node | don't trv node |
     *      |--------------------------------|
     * case | nodeCasePaths | otherCasePaths |
     * ctrl | nodeCtrlPaths | otherCtrlPaths |
     */
    public double fisherExactP(Node n) {
        Map<String,Integer> countsMap = getLabelCounts(n);
        int nodeCasePaths = 0; if (countsMap.containsKey("case")) nodeCasePaths = countsMap.get("case");
        int nodeCtrlPaths = 0; if (countsMap.containsKey("ctrl")) nodeCtrlPaths = countsMap.get("ctrl");
        int otherCasePaths = labelCounts.get("case") - nodeCasePaths;
        int otherCtrlPaths = labelCounts.get("ctrl") - nodeCtrlPaths;
        return fisherExact.getTwoTailedP(nodeCasePaths, otherCasePaths, nodeCtrlPaths, otherCtrlPaths);
    }

    /**
     * Print out the nodes
     * public long id;
     * public String rs;
     * public String contig;
     * public int start;
     * public int end;
     * public String genotype;
     */
    public void printNodes(PrintStream out) {
        if (out==System.out) printHeading("NODES");
        for (Node n : getNodes()) {
            out.println(n.id+"\t"+n.rs+"\t"+n.contig+"\t"+n.start+"\t"+n.end+"\t"+n.genotype+"\t"+n.gf);
        }
    }

    /**
     * Print the paths, labeled by path.getName().
     */
    public void printPaths(PrintStream out) {
        if (out==System.out) printHeading("PATHS");
        for (Path path : paths) {
	    out.println(path.toString());
        }
    }

    /**
     * Print a delineating heading, for general use.
     */
    static void printHeading(String heading) {
        for (int i=0; i<heading.length(); i++) System.out.print("="); System.out.println("");
        System.out.println(heading);
        for (int i=0; i<heading.length(); i++) System.out.print("="); System.out.println("");
    }

    /**
     * Load the graph nodes from a nodes.txt file without filtering.
     */
    public void loadNodesFromTXT(File nodesFile) throws IOException {
        TXTImporter importer = new TXTImporter(nodesFile, null);
	importer.verbose = verbose;
        importer.readNodes();
	buildGraphNodes(importer.nodes);
    }

    /**
     * Load the graph paths from a paths.txt file without any constraints.
     */
    public void loadPathsFromTXT(File nodesFile, File pathsFile) throws IOException {
	loadPathsFromTXT(nodesFile, pathsFile, 0, false);
    }

    /**
     * Load all graph paths from a paths.txt file with the given constraints.
     */
    public void loadPathsFromTXT(File nodesFile, File pathsFile, int maxCases, boolean equalizeCasesControls) throws IOException {
        TXTImporter importer = new TXTImporter(nodesFile, pathsFile);
	importer.verbose = verbose;
	importer.readNodes();
        importer.readPaths(nodeIdMap);
	buildGraphNodes(importer.nodes);
        buildGraphPaths(importer.sampleNodeSets, importer.nodeSamples, maxCases, equalizeCasesControls);
    }

    /**
     * Load graph paths for the given samples in labels.txt from a paths.txt file with the given constraints.
     */
    public void loadPathsFromTXT(File labelsFile, File nodesFile, File pathsFile, int maxCases, boolean equalizeCasesControls) throws IOException {
        TXTImporter importer = new TXTImporter(nodesFile, pathsFile);
	importer.verbose = verbose;
	importer.readNodes();
        importer.readPaths(labelsFile, nodeIdMap);
	buildGraphNodes(importer.nodes);
        buildGraphPaths(importer.sampleNodeSets, importer.nodeSamples, maxCases, equalizeCasesControls);
    }

    /**
     * Load the graph nodes from a VCF file.
     */
    public void loadNodesFromVCF(File vcfFile, double minMAF, double maxMAF) throws IOException {
        VCFImporter importer = new VCFImporter(vcfFile);
	importer.verbose = verbose; 
        importer.readNodes(minMAF, maxMAF);
	buildGraphNodes(importer.nodes);
    }

    /**
     * Load the graph nodes and paths from a VCF file and labels file.
     */
    public void loadPathsFromVCF(File vcfFile, File labelsFile, double minMAF, double maxMAF, int maxCases, boolean equalizeCasesControls) throws IOException {
        VCFImporter importer = new VCFImporter(vcfFile);
	importer.verbose = verbose;
	importer.readNodes(minMAF, maxMAF);
        importer.readPaths(labelsFile, nodeIdMap);
	buildGraphNodes(importer.nodes);
	buildGraphPaths(importer.sampleNodeSets, importer.nodeSamples, maxCases, equalizeCasesControls);
    }

    /**
     * Load the graph nodes from a PLINK list output file.
     * TODO: add filtering!
     */
    public void loadNodesFromList(File listFile) throws IOException {
        ListImporter importer = new ListImporter(listFile);
	importer.verbose = verbose;
        importer.readNodes();
        buildGraphNodes(importer.nodes);
    }

    /**
     * Load the graph paths from a PLINK list output file and labels file
     */
    public void loadPathsFromList(File listFile, File labelsFile, int maxCases, boolean equalizeCasesControls) throws IOException {
	ListImporter importer = new ListImporter(listFile);
	importer.verbose = verbose;
	importer.readNodes();
	importer.readPaths(labelsFile, nodeIdMap);
        buildGraphNodes(importer.nodes);
	buildGraphPaths(importer.sampleNodeSets, importer.nodeSamples, maxCases, equalizeCasesControls);
    }

    /**
     * The main event.
     */
    public static void main(String[] args) throws IOException, FileNotFoundException {
	Options options = new Options();
        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();
        CommandLine cmd;

        Option verboseOption = new Option("v", "verbose", false, "verbose output (false)");
        verboseOption.setRequired(false);
        options.addOption(verboseOption);
        // REQUIRED: graph name
        Option graphOption = new Option("g", "graph", true, "name of graph (required)");
        graphOption.setRequired(true);
        options.addOption(graphOption);
	// ACTION: build graph nodes
	Option nodesOption = new Option("n", "buildnodes", false, "build graph nodes (this or --paths required)");
	nodesOption.setRequired(false);
	options.addOption(nodesOption);
	// ACTION: build graph paths
	Option pathsOption = new Option("p", "buildpaths", false, "build graph paths (this or --nodes required)");
	pathsOption.setRequired(false);
	options.addOption(pathsOption);
        // INPUT: VCF file
        Option vcfFileOption = new Option("vcf", "vcffile", true, "build graph nodes/paths from a tabix-indexed gzipped VCF file");
        vcfFileOption.setRequired(false);
        options.addOption(vcfFileOption);
	// INPUT: PLINK -list file
	Option listFileOption = new Option("list", "listfile", true, "build graph nodes/paths from a PLINK list file");
	listFileOption.setRequired(false);
	options.addOption(listFileOption);
	// INPUT: nodes.txt
	Option nodesFileOption = new Option("nodes", "nodesfile", true, "read graph nodes from a nodes.txt file");
	nodesFileOption.setRequired(false);
	options.addOption(nodesFileOption);
        // INPUT: paths.txt 
        Option pathsFileOption = new Option("paths", "pathsfile", true, "read graph paths from a paths file (<name>.paths.txt)");
        pathsFileOption.setRequired(false);
        options.addOption(pathsFileOption);
	// INPUT: labels.txt
        Option labelsFileOption = new Option("l", "labelsfile", true, "tab-delimited file containing one sample<tab>label pair per line for desired paths");
        labelsFileOption.setRequired(false);
        options.addOption(labelsFileOption);
        // FILTER: reduce cases or controls so they are equal in number
        Option equalizeCasesControlsOption = new Option("ecc", "equalizecasescontrols", false, "reduce cases or controls to make equal in number [false]");
        equalizeCasesControlsOption.setRequired(false);
        options.addOption(equalizeCasesControlsOption);
	// FILTER: set max case paths
	Option maxCasesOption = new Option("maxc", "maxcases", true, "maxiumum number of case paths in the graph [no limit]");
	maxCasesOption.setRequired(false);
	options.addOption(maxCasesOption);
	// FILTER: minMAF
	Option minMAFOption = new Option("minmaf", "minmaf", true, "minimum MAF of loci to be included on graph [0.01]");
	minMAFOption.setRequired(false);
	options.addOption(minMAFOption);
	// FILTER: maxMAF
	Option maxMAFOption = new Option("maxmaf", "maxmaf", true, "maximum MAF of loci to be included on graph [1.00]");
	maxMAFOption.setRequired(false);
	options.addOption(maxMAFOption);
	
        try {
            cmd = parser.parse(options, args);
        } catch (ParseException e) {
            System.err.println(e.getMessage());
            formatter.printHelp("PangenomicGraph", options);
            System.exit(1);
            return;
        }

        // spit out help and exit if nothing supplied
        if (cmd.getOptions().length==0) {
            formatter.printHelp("PangenomicGraph", options);
            System.exit(1);
            return;
        }

	// check that we have a requested action
	if (!cmd.hasOption("buildnodes") && !cmd.hasOption("buildpaths")) {
	    System.err.println("ERROR: you must request building graph nodes (--nodes) or building graph paths (--paths)");
	    System.exit(1);
	}
	boolean buildNodes = cmd.hasOption("buildnodes");
	boolean buildPaths = cmd.hasOption("buildpaths");
	
        // validation
        boolean haveVCF = cmd.hasOption("vcffile");
	boolean haveList = cmd.hasOption("listfile");
	boolean haveNodes = cmd.hasOption("nodesfile");
	boolean havePaths = cmd.hasOption("pathsfile");
	boolean haveLabels = cmd.hasOption("labelsfile");
	if (buildNodes && !haveVCF && !haveList && !haveNodes) {
	    System.err.println("ERROR: to build a graph you must specify an input file with --vcf, --list, or --nodes");
	    System.exit(1);
	}
	if (buildPaths && !haveVCF && !haveList && !haveNodes) {
	    System.err.println("ERROR: to build graph paths you must specify an input file with --vcf, --list, or --nodes");
	    System.exit(1);
	}
	if (buildPaths && !haveLabels) {
	    System.err.println("ERROR: to build graph paths you must specify samples and labels with --labels");
	    System.exit(1);
	}	    

        // our PangenomicGraph
        PangenomicGraph graph = new PangenomicGraph(cmd.getOptionValue("graph"));
        graph.verbose = cmd.hasOption("verbose");

	// filters
	boolean equalizeCasesControls = cmd.hasOption("equalizecasescontrols");
	int maxCases = 0;
	if (cmd.hasOption("maxcases")) maxCases = Integer.parseInt(cmd.getOptionValue("maxcases"));

	if (buildPaths) {
	    // build the graph paths, (which requires building nodes first)
	    if (haveNodes && havePaths) {
		// simply load the graph from an existing nodes.txt file without any filtering
		graph.loadPathsFromTXT(new File(cmd.getOptionValue("nodesfile")), new File(cmd.getOptionValue("pathsfile")), maxCases, equalizeCasesControls);
	    } else if (haveVCF) {
		// load all of the relevant nodes from a tabix-indexed vcf.gz file
		double minMAF = 0.01;
		double maxMAF = 1.00;
		if (cmd.hasOption("minmaf")) minMAF = Double.parseDouble(cmd.getOptionValue("minmaf"));
		if (cmd.hasOption("maxmaf")) maxMAF = Double.parseDouble(cmd.getOptionValue("maxmaf"));
		graph.loadPathsFromVCF(new File(cmd.getOptionValue("vcffile")), new File(cmd.getOptionValue("labelsfile")), minMAF, maxMAF, maxCases, equalizeCasesControls);
	    } else if (haveList) {
		// load nodes from a plink -list file
		// TODO: add filtering!
		graph.loadPathsFromList(new File(cmd.getOptionValue("listfile")), new File(cmd.getOptionValue("labelsfile")), maxCases, equalizeCasesControls);
	    }
	    graph.tallyLabelCounts();
	    System.err.println("Graph has "+graph.vertexSet().size()+" nodes, "+
			       graph.paths.size()+" paths, and "+graph.edgeSet().size()+" edges with "+
			       graph.labelCounts.get("case")+"/"+graph.labelCounts.get("ctrl")+" cases/controls");
	} else if (buildNodes) {
	    // build just the graph nodes
	    if (haveNodes) {
		// simply load the graph from an existing nodes.txt file without any filtering
		graph.loadNodesFromTXT(new File(cmd.getOptionValue("nodesfile")));
	    } else if (haveVCF) {
		// load all of the relevant nodes from a tabix-indexed vcf.gz file
		double minMAF = 0.0;
		double maxMAF = 1.0;
		if (cmd.hasOption("minmaf")) minMAF = Double.parseDouble(cmd.getOptionValue("minmaf"));
		if (cmd.hasOption("maxmaf")) maxMAF = Double.parseDouble(cmd.getOptionValue("maxmaf"));
		graph.loadNodesFromVCF(new File(cmd.getOptionValue("vcffile")), minMAF, maxMAF);
	    } else if (haveList) {
		// load nodes from a plink -list file
		// TODO: add filtering!
		graph.loadNodesFromList(new File(cmd.getOptionValue("listfile")));
	    }
	}

	// output
	if (!cmd.hasOption("nodesfile") && graph.getNodes().size()>0) {
	    if (graph.verbose) System.err.println("Writing "+graph.getNodesFilename());
	    graph.printNodes(new PrintStream(graph.getNodesFilename()));
	}
	if (!cmd.hasOption("pathsfile") && graph.paths.size()>0) {
	    if (graph.verbose) System.err.println("Writing "+graph.getPathsFilename());
	    graph.printPaths(new PrintStream(graph.getPathsFilename()));
	}
    }

    /**
     * Return the name of the nodes.txt file.
     */
    public String getNodesFilename() {
	return name+".nodes.txt";
    }

    /**
     * Return the name of the paths.txt file.
     */
    public String getPathsFilename() {
	return name+".paths.txt";
    }

    /**
     * Return the nodes.txt file.
     */
    public File getNodesFile() {
	return new File(getNodesFilename());
    }

    /**
     * Return the paths.txt file.
     */
    public File getPathsFile() {
	return new File(getPathsFilename());
    }
}
